### ISSUE TITLE
dnf_package doesn't handle multiarch properly

### ISSUE THREAD

**Issue by jaymzh** [chef/chef#6653]
## Description

Given two arches, it gets confused

Example:

```
dnf_package 'glib2' do
  package_name ['glib2'] * 2
  arch ['x86_64', 'i686']
  action :upgrade
end
```

What you will find is that it will always try to install glib2-${version}.x86_64 and glib2-${version}.i686 because it compares both to the string it gets back from the helper which doesn't include the arch.

Output:

```
[2017-12-08T17:32:33-08:00] INFO: Processing dnf_package[glib2] action install (... line 47)
[2017-12-08T17:32:33-08:00] DEBUG: Providers for generic dnf_package resource enabled on node include: [Chef::Provider::Package::Dnf]
[2017-12-08T17:32:33-08:00] DEBUG: Provider for action install on resource dnf_package[glib2] is Chef::Provider::Package::Dnf
[2017-12-08T17:32:35-08:00] DEBUG: sending '{"action":"whatinstalled","provides":"glib2","release":"4.el7","version":"2.46.2","arch":"x86_64"}' to python helper
[2017-12-08T17:32:36-08:00] DEBUG: got 'glib2 0:2.46.2-4.el7 x86_64' from python helper
[2017-12-08T17:32:36-08:00] DEBUG: parsed glib2-0:2.46.2-4.el7.x86_64 from python helper
[2017-12-08T17:32:36-08:00] DEBUG: discarding output on stderr from python helper: No handlers could be found for logger "dnf"
[2017-12-08T17:32:36-08:00] DEBUG: sending '{"action":"whatinstalled","provides":"glib2","release":"4.el7","version":"2.46.2","arch":"i686"}' to python helper
[2017-12-08T17:32:36-08:00] DEBUG: got 'glib2 nil nil' from python helper
[2017-12-08T17:32:36-08:00] DEBUG: parsed glib2-. from python helper
[2017-12-08T17:32:36-08:00] DEBUG: sending '{"action":"whatavailable","provides":"glib2","release":"4.el7","version":"2.46.2","arch":"x86_64"}' to python helper
[2017-12-08T17:32:36-08:00] DEBUG: got 'glib2 0:2.46.2-4.el7 x86_64' from python helper
[2017-12-08T17:32:36-08:00] DEBUG: parsed glib2-0:2.46.2-4.el7.x86_64 from python helper
[2017-12-08T17:32:36-08:00] DEBUG: sending '{"action":"whatavailable","provides":"glib2","release":"4.el7","version":"2.46.2","arch":"i686"}' to python helper
[2017-12-08T17:32:36-08:00] DEBUG: got 'glib2 0:2.46.2-4.el7 i686' from python helper
[2017-12-08T17:32:36-08:00] DEBUG: parsed glib2-0:2.46.2-4.el7.i686 from python helper
[2017-12-08T17:32:36-08:00] DEBUG: dnf_package[glib2] glib2 0:2.46.2-4.el7.x86_64 needs updating to 2.46.2-4.el7
[2017-12-08T17:32:36-08:00] DEBUG: dnf_package[glib2] glib2  needs updating to 2.46.2-4.el7
[2017-12-08T17:32:56-08:00] INFO: dnf_package[glib2] installed ["glib2", "glib2"] at ["2.46.2-4.el7", "2.46.2-4.el7"]
```

Note here: `glib2 0:2.46.2-4.el7.x86_64 needs updating to 2.46.2-4.el7`

## Chef Version

12.21.16

## Platform Version

CentOS7

## Replication Case
See above

## Client Output
See above
### NOTE: CHEF CLIENT BUGS ONLY

This issue tracker is for the code contained within this repo -- `chef-client`, base `knife` functionality (not
plugins), `chef-apply`, `chef-solo`, `chef-client -z`, etc.

* Requests for new or alternative functionality should be made to [feedback.chef.io](https://feedback.chef.io/forums/301644-chef-product-feedback/category/110832-chef-client)
* [Chef Server issues](https://github.com/chef/chef-server/issues/new)
* [ChefDK issues](https://github.com/chef/chef-dk/issues/new)
* Cookbook Issues (see the https://github.com/chef-cookbooks repos or search [Supermarket](https://supermarket.chef.io) or GitHub/Google)

**Comment by lamont-granquist** [comment]
so x86_64 is installed and i686 isn't, but it still thinks that " glib2 0:2.46.2-4.el7.x86_64 needs updating to 2.46.2-4.el7"

**Comment by jaymzh** [comment]
Yup. And on the next run it'll think both need to be upgraded even though they don't.

**Comment by lamont-granquist** [comment]
this is fixed in the yum provider rewrite in #6540 

[root@default-centos-72]/home/vagrant/chef# rpm -e glib2.i686
[root@default-centos-72]/home/vagrant/chef# bundle exec ./multiarch.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * yum_package[glib2] action upgrade
    - upgrade package glib2 from uninstalled to 0:2.50.3-3.el7.i686
[root@default-centos-72]/home/vagrant/chef# bundle exec ./multiarch.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * yum_package[glib2] action upgrade (up to date)
[root@default-centos-72]/home/vagrant/chef#

need to backport all the fixes there back into the dnf provider code.

**Comment by lamont-granquist** [comment]
closing as it needs to get fixed in the process of #9013
